hoomd/state.py

```diff
@@ -157,21 +159,26 @@ def special_pair_types(self):
 
     @property
     def box(self):
+        """The state's box (a :py:class:`hoomd.box.Box` object).
+
+        Ediing the box directly is not allowed.  For example
+        ``state.box.scale(1.1)`` would not scale the state's box. To set the
+        state's box to a new box ``state.box = new_box`` must be used.
+        """
         b = self._cpp_sys_def.getParticleData().getGlobalBox()
-        L = b.getL()
-        return boxdim(Lx=L.x, Ly=L.y, Lz=L.z,
-                      xy=b.getTiltFactorXY(),
-                      xz=b.getTiltFactorXZ(),
-                      yz=b.getTiltFactorYZ(),
-                      dimensions=self._cpp_sys_def.getNDimensions())
-
-    # Set the system box
-    # \param value The new boundaries (a data.boxdim object)
+        return Box._from_cpp(b)
+
     @box.setter
     def box(self, value):
-        if not isinstance(value, boxdim):
-            raise TypeError('box must be a data.boxdim object')
-        self._cpp_sys_def.getParticleData().setGlobalBox(value._getBoxDim())
+        if not isinstance(value, Box):
+            raise TypeError('box must be a hoomd.box.Box object')
+        if value.dimensions != self._cpp_sys_def.getNDimensions():
+            self._simulation.device.cpp_msg.warning(
+                "Box changing dimensions from {} to {}."
+                "".format(self._cpp_sys_def.getNDimensions(),
+                          value.dimensions))
+            self._cpp_sys_def.setNDimensions(value.dimensions)
```

Comment:
I am not sure. I just didn't see a reason to forbid it.

Reply:
@joaander thoughts?

Reply:
That would just have the same problems as just setting the box to 2D using the `State.box` property, right? It would invalidate positions potentially, but not any more so than just setting it normally. If this setting would cause attempts to access the same data twice then it should just raise an Exception.